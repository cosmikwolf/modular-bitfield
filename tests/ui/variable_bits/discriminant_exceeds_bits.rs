use modular_bitfield::prelude::*;

#[derive(Specifier)]
#[discriminant_bits = 2]  // Only 2 bits allows values 0-3
#[bits = (8, 16, 24, 32)]  // 4 variants means discriminant 3 is max
pub enum ValidEnum {
    A(u8),
    B(u16),
    C(B24),
    D(u32),  // This has implicit discriminant 3, which fits in 2 bits
}

#[derive(Specifier)]
#[discriminant_bits = 2]  // Only 2 bits allows values 0-3
#[bits = (8, 16, 24, 32, 40)]  // 5 variants means discriminant 4 is needed
pub enum TooManyVariants {
    A(u8),
    B(u16),
    C(B24),
    D(u32),
    E(B40),  // This has implicit discriminant 4, which doesn't fit in 2 bits
}

#[derive(Specifier)]
#[discriminant_bits = 3]  // 3 bits allows values 0-7
#[bits = (8, 16)]
pub enum ExplicitDiscriminantTooLarge {
    A(u8),
    #[discriminant = 8]  // 8 doesn't fit in 3 bits
    B(u16),
}

#[derive(Specifier)]
#[discriminant_bits = 1]  // Only 1 bit allows values 0-1
#[bits = (8, 16, 24)]
pub enum MixedDiscriminants {
    A(u8),  // implicit discriminant 0
    #[discriminant = 1]
    B(u16),  // explicit discriminant 1
    C(B24),  // implicit discriminant 2, doesn't fit in 1 bit
}

fn main() {}