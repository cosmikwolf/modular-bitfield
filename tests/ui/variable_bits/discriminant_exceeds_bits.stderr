error: discriminant value 4 for variant E exceeds maximum value 3 for 2 discriminant bits
  --> tests/ui/variable_bits/discriminant_exceeds_bits.rs:21:5
   |
21 |     E(B40),  // This has implicit discriminant 4, which doesn't fit in 2 bits
   |     ^^^^^^

error: discriminant value 8 for variant B exceeds maximum value 7 for 3 discriminant bits
  --> tests/ui/variable_bits/discriminant_exceeds_bits.rs:29:5
   |
29 | /     #[discriminant = 8]  // 8 doesn't fit in 3 bits
30 | |     B(u16),
   | |__________^

error: discriminant value 2 for variant C exceeds maximum value 1 for 1 discriminant bits
  --> tests/ui/variable_bits/discriminant_exceeds_bits.rs:40:5
   |
40 |     C(B24),  // implicit discriminant 2, doesn't fit in 1 bit
   |     ^^^^^^

error: attribute value must be a literal
 --> tests/ui/variable_bits/discriminant_exceeds_bits.rs:5:10
  |
5 | #[bits = (8, 16, 24, 32)]  // 4 variants means discriminant 3 is max
  |          ^^^^^^^^^^^^^^^

error: attribute value must be a literal
  --> tests/ui/variable_bits/discriminant_exceeds_bits.rs:15:10
   |
15 | #[bits = (8, 16, 24, 32, 40)]  // 5 variants means discriminant 4 is needed
   |          ^^^^^^^^^^^^^^^^^^^

error: attribute value must be a literal
  --> tests/ui/variable_bits/discriminant_exceeds_bits.rs:26:10
   |
26 | #[bits = (8, 16)]
   |          ^^^^^^^

error: attribute value must be a literal
  --> tests/ui/variable_bits/discriminant_exceeds_bits.rs:35:10
   |
35 | #[bits = (8, 16, 24)]
   |          ^^^^^^^^^^^

error[E0308]: mismatched types
 --> tests/ui/variable_bits/discriminant_exceeds_bits.rs:9:5
  |
9 |     C(B24),
  |     ^^^^^^
  |     |
  |     expected `u32`, found `B24`
  |     arguments to this function are incorrect
  |
note: associated function defined here
 --> src/lib.rs
  |
  |     fn into_bytes(input: Self::InOut) -> Result<Self::Bytes, OutOfBounds>;
  |        ^^^^^^^^^^
